### ISSUE TITLE
[BUG] Owned keyword does not work as expected 🤢🤷‍♂️

### ISSUE THREAD

**Issue by a62690705** [modularml/mojo#2103]
### Bug description

Mojo's ownership system assumes that when ownership of a value is transferred from one variable to another, the value is not actually moved from memory, but rather the memory location of the value is renamed with the name of the new owner. The value remains in place without moving in memory, right? However, contrary to what is expected, the value in the following example moved from its place to a new location, from location 0x7ffdaf8b5240 to location 0x7ffdaf8b5260.

```mojo
@value
struct MyPet:
    var name: String
    var age: Int

    fn __init__(inout self, owned name: String, age: Int):
        print("address of name    	", Pointer.address_of(name), "\n")

        self.name = name ^
        print("address of self.name    	", Pointer.address_of(self.name))
        self.age = age
        print("address of age	        ", Pointer.address_of(self.age))


def main():
    var s = String("mark")
    print("address of s       	", Pointer.address_of(s))

    var my_pet = MyPet(s, 33)


Output

address of s       	 0x7ffdaf8b5240
address of name    	 0x7ffdaf8b5240 

address of self.name     0x7ffdaf8b5260
address of age	         0x7ffdaf8b5258
``` 

### Steps to reproduce

- Include relevant code snippet or link to code that did not work as expected.
- If applicable, add screenshots to help explain the problem.
- If using the Playground, name the pre-existing notebook that failed and the steps that led to failure.
- Include anything else that might help us debug the issue.

### System information

```shell
- What OS did you do install Mojo on ?
- Provide version information for Mojo by pasting the output of `mojo -v`   24.2
- Provide Modular CLI version by pasting the output of `modular -v`
```

**Comment by soraros** [comment]
The problem is that you did not dig deep enough. `String` is backed by a `List`; `String.__moveinit__` moves the underlying buffer; `List.__moveinit__` reassigns the pointer.

```mojo
@value
struct MyPet:
    var name: String

    fn __init__(inout self, owned name: String):
        self.name = name^
        print(self.name._buffer.data)

def main():
    var s = String("mark")
    print(s._buffer.data)

    var my_pet = MyPet(s)
```

**Comment by a62690705** [comment]
What do you mean by moving the buffer?

**Comment by soraros** [comment]
`self.name` is of type `String` and not `Ref[String]`, so a "new" string is stored inline, but we make use of the move semantics by using `__moveinit__` to save the copy.

`String` is backed by a `List`:
https://github.com/modularml/mojo/blob/4d0b45f74fbcf883c63603a762b0c5c3a11895ed/stdlib/src/builtin/string.mojo#L271-L276

Moving the underlying buffer of a `String`, which calls `List.__moveinit__`:
https://github.com/modularml/mojo/blob/4d0b45f74fbcf883c63603a762b0c5c3a11895ed/stdlib/src/builtin/string.mojo#L379-L385

`List` where `data` points to the actual data:
https://github.com/modularml/mojo/blob/4d0b45f74fbcf883c63603a762b0c5c3a11895ed/stdlib/src/collections/list.mojo#L72-L87

Above mentioned `List.__moveinit__`.
https://github.com/modularml/mojo/blob/4d0b45f74fbcf883c63603a762b0c5c3a11895ed/stdlib/src/collections/list.mojo#L127-L135

**Comment by a62690705** [comment]
Thank you for the clarification. Do you mean the keyword owned There is no malfunction  ?

**Comment by soraros** [comment]
> There is no malfunction

I do think so.

**Comment by LJ-9801** [comment]
So I found something interesting that is related to the post, but correct me if I am wrong. I think the reason why `MyPet struct` still keeps the same address of the `owned name` is because there is no other methods accessing `var s` after the `MyPet(s, 33)` is called. The compiler just optimizes away the memcpy to a new address .  

> The problem is that you did not dig deep enough. `String` is backed by a `List`; `String.__moveinit__` moves the underlying buffer; `List.__moveinit__` reassigns the pointer.

These makes sense if we are only transfering owner ship of a value via a `^`, but in the context of a regular `__init__` with a `owned` argument, the struct should make another copy of the value and gain ownership to that.

I did a little experiment. If I simply add another line underneath the `MyPet(s, 33)`, the address will be different again.

```mojo
@value
struct MyPet:
    var name: String
    var age: Int

    fn __init__(inout self, owned name: String, age: Int):
        print("address of name    	", Pointer.address_of(name), "\n")

        self.name = name ^
        print("address of self.name    	", Pointer.address_of(self.name))
        self.age = age
        print("address of age	        ", Pointer.address_of(self.age))


def main():
    var s = String("mark")
    print("address of s       	", Pointer.address_of(s))

    var my_pet = MyPet(s, 33)

    print(s)
```

**Output**
```
address of s       	 0x7ffcae2d19f0
address of name    	 0x7ffcae2d1a08 

address of self.name    	 0x7ffcae2d1a50
address of age	         0x7ffcae2d1a48
mark
```

**Comment by a62690705** [comment]
```mojo
@value
struct MyPet:
    var name: String
    var age: Int

    fn __init__(inout self, owned name: String, age: Int):
        print("address of name    	", Pointer.address_of(name), "\n")

        self.name = name ^
        print("address of self.name    	", Pointer.address_of(self.name))
        self.age = age
        print("address of age	        ", Pointer.address_of(self.age))


def main():
    var s = String("mark")
    print("address of s       	", Pointer.address_of(s))

    var my_pet = MyPet(s ^, 33)

Output

address of s       	 0x7fff32684770
address of name    	 0x7fff32684770 

address of self.name     0x7fff32684790
address of age	         0x7fff32684788
```

**Comment by soraros** [comment]
@LJ-9801 You are correct in that the fact `s` is not used after the constructor call plays a role; you are incorrect in asserting it "should" make a copy, ownership transferring can happen in two different ways for a `Movable` *and* `Copyable` type like `String` ([ref](https://docs.modular.com/mojo/manual/values/ownership#transfer-implementation-details)).
- Argument is `owned`, `s` is `Copyable`, typechecks, *might* make a copy (or, call `__copyinit__`)
- No use of `s` after initialisation, `s` destroyed immediately ([scenario 2](https://docs.modular.com/mojo/manual/values/ownership#transfer-arguments-owned-and-))
- The lifetime of `s` actually ends, *and* `s` is `Movable`, `__moveinit__` is called as explained above ([ref](https://docs.modular.com/mojo/manual/lifecycle/life#value-decorator))

@a62690705 We did transfer `s` to `my_pet`, the *actual data* is not moved as demonstrated in my code snippet above: `self.name._buffer.data` (the memory address of "mark" on the heap) stays the same. `Pointer.address_of(self.name)` gives you the address of the field holding a string struct, not the address of the *data* of the string. You were looking at the wrong thing. There is no "unexplained error"; actually the [doc](https://docs.modular.com/mojo/manual/lifecycle/life#move-constructor) made it very clear:

> That's generally known as a move operation: the memory block holding the data remains the same (the memory does not actually move), but the pointer to that memory moves to a new variable.

**Comment by soraros** [comment]
Let me try again.

Your mental model (maybe?):
```mojo
# struct MyPet
0 field Int
1 field (???)  # nothing makes sense here
```

Actually, I don't think printing `Pointer.address_of(self.name)` makes sense: it gives the address to the pointer at best, what we want is the *content* of the pointer.

In fact (modulo field ordering):
```mojo
# struct MyPet
0 field Int
# struct String: name
# struct List[UInt8]: _buffer
1 field Pointer[String]: data
  ...
n ...
```
We print the content of the pointer to actual memory (`self.name._buffer.data`), which is not changed.

**Comment by a62690705** [comment]
Are you an employee at modular

**Comment by a62690705** [comment]
Thanks for the clarification

**Comment by soraros** [comment]
> Are you an employee at modular

I don't know how that's relevant, but for the record, I'm not.
